### ISSUE TITLE
fft2 ignores overwrite_x=True

### ISSUE THREAD

**Issue by mmlanger** [scipy/scipy#8500]
The overwrite_x keyword actually has no effect if one uses the convenience function for the 2D fft.
It seems the normal fft (1D) is not affected, I did not test ist for fftn.

Also, the docs could be irritating because the keyword is called 'overwrite_x' and in the description is written "contents of x can be destroyed".

### Reproducing code example:
```
import numpy as np
from scipy.fftpack import fft2

x = np.ones((4, 6), dtype=np.complex128)
print("state before inplace fft2:\n", x)

print("FFT:\n", fft2(x))

fft2(x, overwrite_x=True)
print("state after inplace fft2:\n", x)
```
Output
```
state before inplace fft2:
 [[ 1.+0.j  1.+0.j  1.+0.j  1.+0.j  1.+0.j  1.+0.j]
 [ 1.+0.j  1.+0.j  1.+0.j  1.+0.j  1.+0.j  1.+0.j]
 [ 1.+0.j  1.+0.j  1.+0.j  1.+0.j  1.+0.j  1.+0.j]
 [ 1.+0.j  1.+0.j  1.+0.j  1.+0.j  1.+0.j  1.+0.j]]
FFT:
 [[ 24.+0.j   0.+0.j   0.+0.j   0.+0.j   0.+0.j   0.+0.j]
 [  0.+0.j   0.+0.j   0.+0.j   0.+0.j   0.+0.j   0.+0.j]
 [  0.+0.j   0.+0.j   0.+0.j   0.+0.j   0.+0.j   0.+0.j]
 [  0.+0.j   0.+0.j   0.+0.j   0.+0.j   0.+0.j   0.+0.j]]
state after inplace fft2:
 [[ 1.+0.j  1.+0.j  1.+0.j  1.+0.j  1.+0.j  1.+0.j]
 [ 1.+0.j  1.+0.j  1.+0.j  1.+0.j  1.+0.j  1.+0.j]
 [ 1.+0.j  1.+0.j  1.+0.j  1.+0.j  1.+0.j  1.+0.j]
 [ 1.+0.j  1.+0.j  1.+0.j  1.+0.j  1.+0.j  1.+0.j]]
```

### Scipy/Numpy/Python version information:
```
python  3.6.4
numpy  1.13.3
scipy  1.0.0 
```

**Comment by pv** [comment]
The overwrite_* parameters in all routines mean that the output may be 
rewritten, in case it is useful for performance, but it is not a 
guarantee that this occurs.

**Comment by sudheerachary** [comment]
i want to take up this issue can some one please guide me how to start and i have experience in python & numpy.

**Comment by sudheerachary** [comment]
does this mean `fft2(x, overwrite_x=True)` should overwrite the `x` with result of `fft2(x)`.

**Comment by ilayn** [comment]
Yes, but unfortunately this is by design. If there is a up/down casting required then the intermediate arrays are used and those are the ones that get overwritten. Hence it won't be possible to fix this unless the whole fft library is rewritten.

**Comment by mmlanger** [comment]
Ah I see, it is still kind of strange as the behavior is not really intuitive (an argument that may or may not cause your input to be overwritten?), numpy won't give you such an argument:
https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.fft.fft2.html#numpy.fft.fft2

The solution could maybe to just drop the parameter until a more flexible fft library is available?

**Comment by ilayn** [comment]
I'm not sure if all cases are supported via NumPy too but that's because NumPy uses the C bindings of `fftpack`, that is `fftpack_lite` , similar to LAPACK handling. SciPy uses directly the Fortran source. We have recently added a note at http://scipy.github.io/devdocs/linalg.lapack.html#module-scipy.linalg.lapack and same applies here. It's a datatype dependent overwrite via `f2py` and not related to fft library itself.

**Comment by mmlanger** [comment]
But shouldn't than the input be overwritten if the input array fulfills the requirements?
I thought if I make the array F continuous it should than be overwritten:

```
In [36]: x = np.ones((4, 6), order='F', dtype=np.complex128)

In [37]: fft2(x, overwrite_x=True)
Out[37]: 
array([[24.+0.j,  0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j],
       [ 0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j],
       [ 0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j],
       [ 0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j]])

In [38]: x
Out[38]: 
array([[1.+0.j, 1.+0.j, 1.+0.j, 1.+0.j, 1.+0.j, 1.+0.j],
       [1.+0.j, 1.+0.j, 1.+0.j, 1.+0.j, 1.+0.j, 1.+0.j],
       [1.+0.j, 1.+0.j, 1.+0.j, 1.+0.j, 1.+0.j, 1.+0.j],
       [1.+0.j, 1.+0.j, 1.+0.j, 1.+0.j, 1.+0.j, 1.+0.j]])
```

**Comment by pv** [comment]
Whether the array gets overwritten or not with overwrite_xxx=True is
implementation detail in most of the functions. In particular, the f2py
guarantees don't apply here as you find out.

**Comment by pv** [comment]
Clearly, these flags would best not have been added in the first place
as the performance advantage is not that clear.

However, they're been in the API close to 20 years, and removing them
is again a question of breaking old code vs. cosmetic reasons.

**Comment by mmlanger** [comment]
This is of course true, maybe an explicit hint in the documentation would than be in order to not use/avoid this flags?
I teach sometimes classes in python and when I show them a page of the scipy docs and have to immediately tell them "but well don't use the overwrite keywords or at least test them beforehand" makes the strong advertisement I'm doing beforehand for numpy+scipy somewhat awkward.

**Comment by pv** [comment]
The description of these parameters in the documentation can of course
be made more clear on that they exist only for performance reasons, and
will not necessarily result to overwriting.

**Comment by sudheerachary** [comment]
will the following changes be made to documentation ? if so i would like to take up the issue.
